docs/mount-change-detection.md

```diff
@@ -0,0 +1,58 @@
+# Mount-point change detection
+
+## Introduction
+
+This document presents the design of the mount-point change detection system in NDM.
+The goal of mount-point change detection is to detect the changes in the mount-points
+and the filesystem on the existing blockdevices discovered by NDM and trigger appropriate
+action to update the blockdevice CRs.
+
+## Design
+
+The mount-points and filesystem info of a block device are found by the probe - *mountprobe*.
+This probe uses the mounts file, which the Linux kernel provides in the *procfs* pseudo
+filesystem. Reading the mounts file provides the status of all the mounted
+filesystems (sample output below).
+
+``` text
+rootfs / rootfs rw 0 0 
+/dev/root / ext3 rw 0 0 
+/proc /proc proc rw 0 0 usbdevfs 
+/proc/bus/usb usbdevfs rw 0 0 
+/dev/sda1 /boot ext3 rw 0 0 none 
+/dev/pts devpts rw 0 0 
+/dev/sda4 /home ext3 rw 0 0 none 
+/dev/shm tmpfs rw 0 0 none 
+/proc/sys/fs/binfmt_misc binfmt_misc rw 0 0
+```
+
+Whenever a block device is (un)mounted or the fs type changes, the changes are reflected in the mounts file. This proposal introduces a new probe in NDM called _**mount-change-probe (MCP)**_. The probe will watch for changes in the mounts file and trigger updation when a change is detected. To watch the mounts file for changes, MCP uses the *epoll* API. The *epoll* API is provided by the Linux kernel for the userspace programs to monitor file descriptors and get notifications about I/O events happening on them. Whenever the mounts file changes, the events `EPOLLPRI` and `EPOLLERR` are emitted. This behaviour has been documented [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=31b07093c44a7a442394d44423e21d783f5523b8) (additional links - [\[1\]](https://lkml.org/lkml/2006/2/22/169), [\[2\]](http://lkml.iu.edu/hypermail/linux/kernel/1012.1/02246.html)).
```

Comment:
When the fs type changes, there is no change in the mounts file. `mounts` file reflect only about changes of devices that are mounted.

You can use the [`Start()`](https://github.com/openebs/node-disk-manager/blob/master/cmd/ndm_daemonset/probe/mountprobe.go#L89) in mountprobe itself right? instead of a new probe? Similar way how udevprobe does monitoring?

Reply:
> When the fs type changes, there is no change in the mounts file. `mounts` file reflect only about changes of devices that are mounted.

👍🏽 will add note

> You can use the [`Start()`](https://github.com/openebs/node-disk-manager/blob/master/cmd/ndm_daemonset/probe/mountprobe.go#L89) in mountprobe itself right? instead of a new probe? Similar way how udevprobe does monitoring?

Ture. This looks like the cleaner way to do it.